Index: dispatch/source.h
===================================================================
--- dispatch/source.h	(revision 22)
+++ dispatch/source.h	(working copy)
@@ -30,7 +30,7 @@
 #include <mach/port.h>
 #include <mach/message.h>
 #endif
-#include <sys/signal.h>
+#include <signal.h>
 
 /*!
  * @header
Index: src/internal.h
===================================================================
--- src/internal.h	(revision 22)
+++ src/internal.h	(working copy)
@@ -138,7 +138,7 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <limits.h>
-#include <search.h>
+//#include <search.h>
 #if USE_POSIX_SEM
 #include <semaphore.h>
 #endif
Index: src/shims/getprogname.h
===================================================================
--- src/shims/getprogname.h	(revision 22)
+++ src/shims/getprogname.h	(working copy)
@@ -27,9 +27,11 @@
 static inline const char *
 getprogname(void)
 {
-# if HAVE_DECL_PROGRAM_INVOCATION_SHORT_NAME
+# if defined(ANDROID)
+	return "FIXME -- TODO";
+# elif HAVE_DECL_PROGRAM_INVOCATION_SHORT_NAME
     return program_invocation_short_name;
-#elif HAVE_GETEXECNAME
+# elif HAVE_GETEXECNAME
     return getexecname();
 # else
 #   error getprogname(3) is not available on this platform
